core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.core.data;
+
+
+import org.apache.accumulo.core.security.ColumnVisibility;
+import org.apache.hadoop.io.Text;
+
+/**
+ * A builder used to build {@link Key}s by defining their components.
+ *
+ * The rules are:
+ * <ul>
+ *   <li>All components of the {@link Key} are optional except the row</li>
+ *   <li>Components not explicitly set default to empty byte array except the timestamp which
+ *   defaults to <code>Long.MAX_VALUE</code></li>
+ *   <li>The column qualifier can only be set if the column family has been set first</li>
+ *   <li>The column visibility can only be set if at least the column family has been set first</li>
+ * </ul>
+ *
+ * The builder supports three types of components: <code>byte[]</code>, <code>Text</code> and <code>CharSequence</code>.
+ * <code>CharSequence</code>s must be UTF-8 encoded.
+ *
+ * The builder is mutable and not thread safe.
+ *
+ * @see org.apache.accumulo.core.data.Key
+ * @since 1.9
+ */
+public class KeyBuilder {
+
+  /**
+   * @since 1.9
+   */
+  public interface Build {
+
+    /**
+     * Build a {@link Key} from this builder. copyBytes defaults to true.
+     *
+     * @return
+     *          the {@link Key} built from this builder
+     */
+    Key build();
+
+    /**
+     * Change the timestamp of the {@link Key} created.
+     *
+     * @param timestamp
+     *          the timestamp to use for the {@link Key}
+     * @return this builder
+     */
+    Build timestamp(long timestamp);
+
+    /**
+     * Set the deleted marker of the {@link Key} to the parameter.
+     *
+     * @param deleted
+     *          if the {@link Key} should be marked as deleted or not
+     * @return this builder
+     */
+    Build deleted(boolean deleted);
+  }
+
+  /**
+   * @since 1.9
```

Comment:
We should probably have some sort of javadoc description for these interfaces (RowStep, ColumnFamilyStep, etc) since they are in the public API. One sentence should be sufficient.